### ISSUE TITLE
Compiler (gcc 10.1.0) warning in: /src/cmd/beam_generic.cpp

### ISSUE THREAD

**Issue by evertvorster** [vegastrike/Vega-Strike-Engine-Source#227]
```
[  4%] Building CXX object CMakeFiles/engine_com.dir/src/cmd/beam_generic.cpp.o
/home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/cmd/beam_generic.cpp: In member function 'void Beam::Init(const Transformation&, const weapon_info&, void*, Unit*)':
/home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/cmd/beam_generic.cpp:95:51: warning: 'void* memset(void*, int, size_t)' clearing an object of non-trivial type 'struct GFXColorVertex'; use assignment or value-initialization instead [-Wclass-memaccess]
   95 |         memset( beam, 0, sizeof (*beam)*numvertex );
      |                                                   ^
In file included from /home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/star_system_generic.h:7,
                 from /home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/universe_generic.h:28,
                 from /home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/vs_globals.h:5,
                 from /home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/gfx/matrix.h:26,
                 from /home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/gfx/quaternion.h:5,
                 from /home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/gfx/mesh.h:27,
                 from /home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/cmd/beam.h:4,
                 from /home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/cmd/beam_generic.cpp:3:
/home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/gfxlib_struct.h:234:8: note: 'struct GFXColorVertex' declared here
  234 | struct GFXColorVertex
      |        ^~~~~~~~~~~~~~
/home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/cmd/beam_generic.cpp: In member function 'bool Beam::Collide(Unit*, Unit*, Unit*)':
/home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/cmd/beam_generic.cpp:361:14: warning: 'nonnull' argument 'this' compared to NULL [-Wnonnull-compare]
  361 |     if (this == NULL || target == NULL) {
      |              ^
[  4%] Building CXX object CMakeFiles/engine_com.dir/src/cmd/bolt_generic.cpp.o
```

There seems to be two different errors in this file. 
One of those clearing of non-trivial type errors, and comparing a non-NULL to NULL.

**Comment by evertvorster** [comment]
This error is breaking my brain. 
The internet at large condemns using memset in C++, especially to initialize functions. What it does not do is show me a way around this. Any help in explaining what is happening would enable me to clear up the other 5 areas where this is used in our code.

**Comment by evertvorster** [comment]
According to https://stackoverflow.com/questions/51658046/memset-and-a-dynamic-array-of-stdcomplexdouble, answer 3, line 94 of beam_generic.cpp already initializes the beam, so line 95 can just go away. I will be play-testing this shortly to see if that works.

**Comment by stephengtuggy** [comment]
`beam` is an array. It's not very clear, but see the square brackets `[ ]` around `numvertex` on line 94? That's initializing an array with a dynamic size.

So now we need to make sure that each element in the array is initialized. If the `GFXColorVertex` constructor doesn't do it, then we need to do it here, probably using a `for` loop.

**Comment by evertvorster** [comment]
What does the "new" keyword below do? According to that link, it makes a new array, should that not mean that it also initializes the array?
GFXColorVertex *beam = new GFXColorVertex[numvertex]; 
In any case, I have been trying to break this function and then see the breakage in the game to make sure I know what to test, but so far I have been unable to. I'm pretty sure this has something to do with the beams when you are shooting, but this function seems to work no matter what I do to it.

**Comment by stephengtuggy** [comment]
You're right, it does initialize the array also. Or at least, it calls the no-argument `GFXColorVertex` constructor.

I'm coming to the conclusion that the problem is the above constructor. It doesn't initialize the member fields like it should. That's probably what we should fix.

**Comment by stephengtuggy** [comment]
gfxlib_struct.h, line 262.

**Comment by stephengtuggy** [comment]
Something like this:

```C++
    GFXColorVertex() {
        this->x = 0.f;
        this->y = 0.f;
        this->z = 0.f;
        this->i = 0.f;
        this->j = 0.f;
        this->k = 0.f;
        this->r = 0.f;
        this->g = 0.f;
        this->b = 0.f;
        this->a = 0.f;
        this->s = 0.f;
        this->t = 0.f;
        this->tx = 0.f;
        this->ty = 0.f;
        this->tz = 0.f;
        this->tw = 1.f;
    }
```

(I took the default value for tw from one of the other constructors, which uses 1.f as the default value for tw/parity if not specified.)

**Comment by stephengtuggy** [comment]
While we're at it, let's make sure all the other constructors initialize all the fields as well.

**Comment by evertvorster** [comment]
@stephengtuggy This thing is currently way above my understanding. All I know is that there are six instances in our code where memset is used to initialize something that is non-trivial, this being one. 
I thought that staring at it for a day would enlighten me, but so far it's been a very depressing experience, realizing just how little I know of this stuff.

**Comment by evertvorster** [comment]
I'm re-building the source now to see if your suggested changes makes a difference. :)

**Comment by stephengtuggy** [comment]
Did you remove the `memset` line also? (If you don't, you'll still get the warning.)

**Comment by evertvorster** [comment]
I have commented out the memset line. However, I was absolutely sure that this function would affect beamed weapons, but no matter what I do to it, my guns work fine in the game.

**Comment by evertvorster** [comment]
This has been fixed by #275
